### PR TITLE
Fix collections coerce float

### LINKED ISSUE #21803: [Bug]: using `set_offsets` on scatter object raises TypeError
### Bug summary

`plt.scatter` can accept datetime as arguments, but `set_offsets` method of it's returns cannot.

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(0, 10)
y = np.sin(x)
d = x * np.timedelta64(24, 'h') + np.datetime64('2021-11-29')

sc = plt.scatter(d, y) # shows a plot
sc.set_offsets(list(zip(d, y))) # clears the plot so it becomes just white square
```

### Actual outcome

The scatters just disappear if you use other `timedelta64` units than `D`, also if you use other date unit, than day.

### Expected outcome

Excepted outcome is to update positions of scatters

### Additional information

_No response_

### Operating system

Arch

### Matplotlib Version

3.5.0

### Matplotlib Backend

Qt5Agg, `'module://ipympl.backend_nbagg'`

### Python version

Python 3.9.6

### Jupyter version

3.1.7

### Installation

conda

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Collections `set_offset` was coercing to float, which ruins any units we want to do.  Closes #21803

Needs tests....

Tested with 
```python
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(0, 10)
y = np.sin(x)
d = x * np.timedelta64(24, 'h') + np.datetime64('2021-11-29')

sc = plt.scatter(d, y) # showsw a plot
sc.set_offsets(list(zip(d, y)))
plt.show()
```

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -545,7 +545,10 @@ def set_offsets(self, offsets):
         ----------
         offsets : (N, 2) or (2,) array-like
         """
-        offsets = np.asanyarray(offsets, float)
+        offsets = np.asanyarray(offsets)
+        offsets[:, 0] = self.convert_xunits(offsets[:, 0])
+        offsets[:, 1] = self.convert_xunits(offsets[:, 1])
```
Comment (QuLogic) [inline lib/matplotlib/collections.py:550]: ?
```suggestion
        offsets[:, 1] = self.convert_yunits(offsets[:, 1])
```

Comment (jklymak) [reply lib/matplotlib/collections.py]: Ooops, yes, fixed.  The conversion also had to take place after co-ercing to 2xN...

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -545,9 +545,12 @@ def set_offsets(self, offsets):
         ----------
         offsets : (N, 2) or (2,) array-like
         """
-        offsets = np.asanyarray(offsets, float)
+        offsets = np.asanyarray(offsets)
         if offsets.shape == (2,):  # Broadcast (2,) -> (1, 2) but nothing else.
             offsets = offsets[None, :]
+        offsets[:, 0] = self.convert_xunits(offsets[:, 0])
+        offsets[:, 1] = self.convert_yunits(offsets[:, 1])
+        offsets = np.asarray(offsets, 'float')
```
Comment (QuLogic) [inline lib/matplotlib/collections.py:553]: If we're going to re-cast, is there any reason to hang on to the original array? Can we do `np.column_stack((convertx(...), converty(...)))`?

Comment (jklymak) [reply lib/matplotlib/collections.py]: sure, done.  Not sure it is any more elegant, but probably better to convert the x and y arrays separately to float anyway (they can be object arrays, even though convert_xunits has been called, which actually strikes me as a problem in convert_x/yunits.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**